### PR TITLE
Ferrocene Self Test: Check C compiler passes extra args.

### PR DESCRIPTION
**Pull request by jonathanpallant**
We now give the C compiler a `-Wl,--foobar` argument, and test that the linker receives `--foobar` as an argument (but using a long random string instead of `foobar`).

### CODE REVIEW

**Comment by jonathanpallant** [comment]
Now we execute sample programs (if they are executables, and we're on the host target) so now we check the object files are linked correctly.

**Comment by jonathanpallant** [comment]
bors try

File: ferrocene/tools/self-test/src/compile.rs [hunk 1]
```diff
@@ -119,12 +135,35 @@ fn compile(ctx: &Context<'_>, program: &SampleProgram) -> Result<(), Error> {
...
+                name: program.name.into(),
+                error,
+            })?;
+        if output.stdout != expected_output {
+            // TODO change for SampleProgramOutputWrong
+            return Err(Error::SampleProgramOutputWrong {
+                name: program.name.into(),
+                expected: expected_output.to_vec(),
+                found: output.stdout,
+            });
```
Comment (pietroalbini) [inline ferrocene/tools/self-test/src/compile.rs:163]: Why the TODOs here?

**Review by pietroalbini — COMMENTED** [review]
Code changes look good, only left a small comment on some leftover TODOs. Also, we should add a requirement for this in the safety manual (https://public-docs.ferrocene.dev/main/safety-manual/options.html#linker-options)

File: ferrocene/tools/self-test/src/compile.rs (hunk 1 above)
Comment (jonathanpallant) [reply ferrocene/tools/self-test/src/compile.rs:163]: ah, because I fixed the TODO and didn't delete the note I'd left myself

**Comment by jonathanpallant** [comment]
Added a para to the safety manual about what the C compiler must do.

**Pull request merged**